### DISCUSSION TITLE
Alternative for core.setObserverLocation, for spacecraft RA/Dec/distance?

### DISCUSSION THREAD

**Discussion by stephanpls** [Stellarium/stellarium#2108] [Q&A]
With core.setObserverLocation you can change your observer location to another object. That is to say, an object that Stellarium knows, such as Jupiter or 1 Ceres.
My goal is to "put myself" on a spacecraft, let's say BepiColombo, the Mercury explorer, and look back to the Earth-Moon system. 

But Stellarium does not know BepiColombo, and not many, if any, other spacecraft. Sure, I could put its Kepler elements in ssystem_minor.ini, but that's quite laborious every time with changing trajectories.

At ssd.jpl.nasa.gov/horizons I can download the trajectories of many spacecraft in RA, Dec and distance, as seen from the Earth. Right from the beginning of their  trajectory (launch), so you have the whole history in one file. 

I would like to use those RA, Dec and distance as input for something like core.setObserverLocation. Is such an option available?

**Comment by alex-w** [comment]
It’s not implemented yet, see #259 and #448

**Comment by gzotti** [comment]
In a plugin for the far future I have plans to implement spacecraft trajectories. However, RA/DE lists will not help. You need 3D trajectory data in J2000 ecliptical coordinates. Then one could subclass Orbit to read and interpolate such data files and move some space probe (even a 3D model) through the Solar system. However, don't hold your breath, only after urgent tasks have been fixed.

**Comment by stephanpls** [comment]
Thanks. Well, I won't hold my breath then, no problem!
I plan to develop myself a tool from within my MS Access application (that interacts with Stellarium by scripting with ssc-files).
It's my idea that this tool will get hold of spacecraft Kepler elements using the ssd.jpl.nasa.gov/horizons API.
Then it will add the API output to ssystem_minor.ini, and then it should be possible what I intend ("put myself" on a spacecraft, see beginning of this thread).
Nice research and developing for me ahead, big fun!
As almost always, the developing itself gives more fun than finally using the result of it....

**Comment by stephanpls** [comment]
Here the result for Voyager 1 today, the spacecraft most far away from Earth.
I wrote some code to get the osculating Kepler-elements of its orbit via the JPL Horizons API, processed its output and added it then in the right format to ssystem_minor.ini. Nice trial and error work!

The first picture shows how we "see" the Voyager 1 today at more than 23 billion km. Note its magnitude!
In the second picture we see the Earth (almost magnitude 8 with apparent diameter 0.1"!) from the Voyager, and also the nearby Sun at 14', Mercury and Venus.

We could do the same for the about 190 spacecraft available in the JPL Horizons API.

![Voyager 1 from Earth](https://user-images.githubusercontent.com/96252651/147651631-88c47e87-28b2-41e5-9f79-8779f9c1edaa.jpg)
![Earth from Voyager 1](https://user-images.githubusercontent.com/96252651/147651653-b4fb412b-8d9f-4a86-97b1-1011131faf75.jpg)

**Reply by gzotti** [reply]
Nice work!
Note that Voyager's magnitude estimate depends on its diameter, though ;-)
Given that orbital elements don't change except at flybys, you could prepare one file in this format to be loaded with the SSE plugin. The orbit_good time can probably be set to 10 or even 50 years (depends on spacecraft and project intent. Of course it is only good *between* flybys, but we cannot model this currently). We could provide this file for others as preliminary solution until somebody implements a full interplanetary space probe plugin (which visualizes the recorded trajectories from JPL in SPICE format, includes 3D models, and allows the user to observe the spacecraft as it traverses the solar system and rotates as it observes the planets...).

**Comment by stephanpls** [comment]
I thought already about providing the Stellarium community with such an .ini file (for instance called ssystem_spacecraft.ini). I will prepare one with the elements of Pioneer 1 and 2, Voyager 1 and 2 and New Horizons, the spacecraft farthest away (and for me the most interesting). Of course I could add others. 
Ready in the days to come.
